### PR TITLE
Unify max-fee interfaces across commands and cheatcodes

### LINKED ISSUE #935: Unify max-fee interfaces across commands and cheatcodes

### PR DESCRIPTION
**Pull request by karol-bisztyga**
fix #935

### CODE REVIEW

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py [hunk 1]
```diff
@@ -25,14 +25,12 @@
 
 class SignedCheatcodeConfig(CheatcodeNetworkConfig):
     max_fee: NotRequired[Wei]
```
Comment (kasperski95) [inline protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py:27]:
```suggestion
    max_fee: NotRequired[Fee]
```

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py [hunk 2]
```diff
@@ -25,14 +25,12 @@
...
-    auto_estimate_fee: NotRequired[bool]
 
 
 @dataclass
 class ValidatedSignedCheatcodeConfig:
     max_fee: Optional[Wei] = None
```
Comment (kasperski95) [inline protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py:32]:
```suggestion
    max_fee: Optional[Fee] = None
```

File: protostar/starknet_gateway/gateway_facade.py [hunk 3]
```diff
@@ -212,7 +210,7 @@ async def _create_declare_tx_v1(
         )
         # pylint: disable=protected-access
         max_fee = await account_client._get_max_fee(
-            transaction=declare_tx, max_fee=max_fee, auto_estimate=auto_estimate_fee
+            transaction=declare_tx, max_fee=max_fee, auto_estimate=max_fee is None
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade.py:213]:
```suggestion
            transaction=declare_tx, max_fee=max_fee, auto_estimate=max_fee == "auto"
```
max_fee should always be provided to protect from unexpected costs. `max_fee` will be required arguments once we drop support for transactions v0

File: protostar/starknet_gateway/gateway_facade.py [hunk 4]
```diff
@@ -333,7 +329,7 @@ async def invoke(
                 contract_function,
                 inputs,
                 max_fee=max_fee,
-                auto_estimate=auto_estimate_fee,
+                auto_estimate=max_fee is None,
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade.py:332]:
```suggestion
                auto_estimate=max_fee=="auto",
```

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md [hunk 5]
```diff
@@ -19,7 +19,6 @@ It's an extension of [CheatcodeNetworkConfig](../03-network-config.md), so it's
 Wei = int
 class SignedCheatcodeConfig(CheatcodeNetworkConfig):
     max_fee: Wei | None
```
Comment (kasperski95) [inline website/docs/tutorials/06-deploying/02-migrations/invoke.md:21]:
```suggestion
    max_fee: Fee
```

**Review by kasperski95 — CHANGES_REQUESTED** [review]
`max_fee` should be of type `Fee`

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md (hunk 5 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/06-deploying/02-migrations/invoke.md]: [change type of max_fee to Fee](https://github.com/software-mansion/protostar/pull/979/commits/8b3d0ca5d7f7c558635efaccc2bc6c78e0f11c82)

File: protostar/starknet_gateway/gateway_facade.py (hunk 3 above)
Comment (karol-bisztyga) [reply protostar/starknet_gateway/gateway_facade.py]: [change type of max_fee to Fee](https://github.com/software-mansion/protostar/pull/979/commits/8b3d0ca5d7f7c558635efaccc2bc6c78e0f11c82)

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py (hunk 1 above)
Comment (karol-bisztyga) [reply protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py]: [change type of max_fee to Fee](https://github.com/software-mansion/protostar/pull/979/commits/8b3d0ca5d7f7c558635efaccc2bc6c78e0f11c82)

File: protostar/starknet_gateway/gateway_facade.py (hunk 3 above)
Comment (karol-bisztyga) [reply protostar/starknet_gateway/gateway_facade.py]: This is an external call and we have to operate on the type that's there: `max_fee: Optional[int] = None,` or did I miss something?

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py [hunk 6]
```diff
@@ -42,8 +39,7 @@ def from_dict(
             return cls()
         return cls(
             wait_for_acceptance=config.get("wait_for_acceptance", False),
-            max_fee=config.get("max_fee", None),
-            auto_estimate_fee=config.get("auto_estimate_fee", False),
+            max_fee=config.get("max_fee", "auto"),
```
Comment (kasperski95) [inline protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py:42]: If `max_fee` is undefined, Protostar should throw an error to the user.

Note: Currently, `config` parameter is optional. We need to make it required for migration cheatcodes to force the user passing `max_fee`. However, this change will break the cheatcode Protocol between migration and testing module (unless we make it generic). This is a task for a different PR though.

File: protostar/starknet_gateway/gateway_facade.py [hunk 7]
```diff
@@ -305,8 +305,7 @@ async def invoke(
         account_address: str,
         signer: BaseSigner,
         inputs: Optional[CairoOrPythonData] = None,
-        max_fee: Optional[int] = None,
-        auto_estimate_fee: bool = False,
+        max_fee: Fee = "auto",
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade.py:308]:
```suggestion
        max_fee: Fee,
```

**Review by kasperski95 — CHANGES_REQUESTED** [review]

File: protostar/starknet_gateway/gateway_facade.py (hunk 3 above)
Comment (kasperski95) [reply protostar/starknet_gateway/gateway_facade.py]: I'm not sure if I get your point. The `Fee` type can be adapted to Starknet.py interface.

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py [hunk 8]
```diff
@@ -17,22 +17,19 @@
...
 )
 from protostar.starknet.data_transformer import CairoOrPythonData
 
-Wei = int
-
 
 class SignedCheatcodeConfig(CheatcodeNetworkConfig):
-    max_fee: NotRequired[Wei]
-    auto_estimate_fee: NotRequired[bool]
+    max_fee: NotRequired[Fee]
```
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py:26]: You're right

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py [hunk 9]
```diff
@@ -17,22 +17,19 @@
...
 
 
 @dataclass
 class ValidatedSignedCheatcodeConfig:
-    max_fee: Optional[Wei] = None
+    max_fee: Fee = "auto"
```
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py:31]: Why you do not want a default value here?

[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py:31]: It shouldn't be needed if the `max_fee` is required.

File: protostar/starknet_gateway/gateway_facade.py (hunk 3 above)
Comment (karol-bisztyga) [reply protostar/starknet_gateway/gateway_facade.py]: I understand, but in this case, we'd first need it there in the starknet.py as it is an external lib to the protostar, right? And only then update the version in the protostar and update the code here accordingly. Am I correct?

File: protostar/starknet_gateway/gateway_facade.py (hunk 7 above)
Comment (karol-bisztyga) [reply protostar/starknet_gateway/gateway_facade.py]: [make max_fee mandatory and update tests](https://github.com/software-mansion/protostar/pull/979/commits/b2e0a73bd1102a4cb0555cb2e3e91535b785d958)

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py (hunk 6 above)
Comment (karol-bisztyga) [reply protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py]: [make max_fee mandatory and update tests](https://github.com/software-mansion/protostar/pull/979/commits/b2e0a73bd1102a4cb0555cb2e3e91535b785d958)

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py]: [make max_fee mandatory and update tests](https://github.com/software-mansion/protostar/pull/979/commits/b2e0a73bd1102a4cb0555cb2e3e91535b785d958)

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py]: [make max_fee mandatory and update tests](https://github.com/software-mansion/protostar/pull/979/commits/b2e0a73bd1102a4cb0555cb2e3e91535b785d958)

File: protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/migrator/cheatcodes/migrator_invoke_cheatcode.py]: [make max_fee mandatory and update tests](https://github.com/software-mansion/protostar/pull/979/commits/b2e0a73bd1102a4cb0555cb2e3e91535b785d958)

**Review by kasperski95 — APPROVED** [review]

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md [hunk 10]
```diff
@@ -18,8 +18,7 @@ It's an extension of [CheatcodeNetworkConfig](../03-network-config.md), so it's
 ```python
 Wei = int
 class SignedCheatcodeConfig(CheatcodeNetworkConfig):
-    max_fee: Wei | None
-    auto_estimate_fee: int
+    max_fee: Fee
```
Comment (kasperski95) [inline website/docs/tutorials/06-deploying/02-migrations/invoke.md:21]: `Wei` type is not used, and it should be replaced with `Fee`.

**Review by kasperski95 — CHANGES_REQUESTED** [review]

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md [hunk 11]
```diff
@@ -66,7 +63,6 @@ func up() {
             "initialize",
             {"new_authority": 123},
             config={
-                "auto_estimate_fee": True,
                 "wait_for_acceptance": True,
```
Comment (kasperski95) [inline website/docs/tutorials/06-deploying/02-migrations/invoke.md:66]: I think this config should include `max_fee`.

Comment (karol-bisztyga) [reply website/docs/tutorials/06-deploying/02-migrations/invoke.md:66]: [fix docs](https://github.com/software-mansion/protostar/pull/979/commits/794612a37db566139eaf1c36bad0577a0d6e83fc)

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md (hunk 10 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/06-deploying/02-migrations/invoke.md]: [fix docs](https://github.com/software-mansion/protostar/pull/979/commits/794612a37db566139eaf1c36bad0577a0d6e83fc)

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md [hunk 12]
```diff
@@ -16,10 +16,9 @@ This cheatcode invokes a StarkNet contract, with possible state changes. Can be
 It's an extension of [CheatcodeNetworkConfig](../03-network-config.md), so it's properties are applicable here as well.
 
 ```python
-Wei = int
+from protostar.starknet_gateway import Fee
```
Comment (kasperski95) [inline website/docs/tutorials/06-deploying/02-migrations/invoke.md:19]:
```suggestion
Wei = int
Fee = "auto" | Wei
```

File: tests/e2e/test_migrating/migration_signed_invokes.cairo [hunk 13]
```diff
@@ -5,7 +5,7 @@ func up() {
     %{
         contract  = deploy_contract("./build/main.json")
 
-        invoke(contract.contract_address, "put", { "value": 69 }, config={"auto_estimate_fee": True})
+        invoke(contract.contract_address, "put", { "value": 69 }, config={"max_fee": "auto"})
```
Comment (kasperski95) [inline tests/e2e/test_migrating/migration_signed_invokes.cairo:8]: I think Protostar should check if auto_estimate_fee exists. If so, tell the user to use `max_fee="auto"`.

(Btw. In the migration documentation, we have a warning `Breaking changes can be introduced without deprecation.`)

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md [hunk 14]
```diff
@@ -5,8 +5,8 @@ def invoke(
     contract_address: int,
     function_name: str,
     inputs: list[int] | dict[str, Any] | None = None,
+    config: SignedCheatcodeConfig,
     *,
```
Comment (kasperski95) [inline website/docs/tutorials/06-deploying/02-migrations/invoke.md:9]: I think `config` should be below the `*`. It's a keyword only argument.

File: protostar/starknet_gateway/gateway_facade.py (hunk 3 above)
Comment (kasperski95) [reply protostar/starknet_gateway/gateway_facade.py]: I mean, we could make `max_fee` required in `GatewayFacade::_create_declare_tx_v1`. Anyway, it's a private method. Resolving.

Comment (karol-bisztyga) [reply protostar/starknet_gateway/gateway_facade.py]: It is required now in `_create_declare_tx_v1`

File: tests/e2e/test_migrating/migration_signed_invokes.cairo (hunk 13 above)
Comment (karol-bisztyga) [reply tests/e2e/test_migrating/migration_signed_invokes.cairo:8]: [add warning about auto_estimate_fee deprecation](https://github.com/software-mansion/protostar/pull/979/commits/d41edcf97aae14c0562b2c82fbb576e5aa91c62d)

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md (hunk 12 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/06-deploying/02-migrations/invoke.md]: [fix type in docs](https://github.com/software-mansion/protostar/pull/979/commits/701c140ae19de3937f43a7dd8828e489cff4de5b)

File: website/docs/tutorials/06-deploying/02-migrations/invoke.md (hunk 14 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/06-deploying/02-migrations/invoke.md:9]: [reorder args so config is keyword-only](https://github.com/software-mansion/protostar/pull/979/commits/4867cf55bafc6a338417888275d52879bf529e2a)

**Review by kasperski95 — APPROVED** [review]

**Pull request merged**